abi-versions/v0.2.1/README.md

```diff
@@ -0,0 +1,1996 @@
+- `PLUGIN_CONFIGURATION` = `7`
+- `FOREIGN_FUNCTION_ARGUMENTS` = `8`
+
+
+#### `proxy_header_map_type_t`
+
+- `HTTP_REQUEST_HEADERS` = `0`
+- `HTTP_REQUEST_TRAILERS` = `1`
+- `HTTP_RESPONSE_HEADERS` = `2`
+- `HTTP_RESPONSE_TRAILERS` = `3`
+- `GRPC_CALL_INITIAL_METADATA` = `4`
+- `GRPC_CALL_TRAILING_METADATA` = `5`
+- `HTTP_CALL_RESPONSE_HEADERS` = `6`
+- `HTTP_CALL_RESPONSE_TRAILERS` = `7`
+
+
+#### `proxy_peer_type_t`
+
+- `UNKNOWN` = `0`
+- `LOCAL` = `1`
+- `REMOTE` =`2`
+
+
+#### `proxy_stream_type_t`
+
+- `HTTP_REQUEST` = `0`
+- `HTTP_RESPONSE` = `1`
+- `DOWNSTREAM` = `2`
+- `UPSTREAM` = `3`
+
+
+#### `proxy_metric_type_t`
+
+- `COUNTER` = `0`
+- `GAUGE` = `1`
+- `HISTOGRAM` = `2`
+
+
+#### `wasi_errno_t`
+
+- `SUCCESS` = `0`
+- `BADF` = `8`
+- `FAULT` = `21`
+- `INVAL` = `28`
+- `NOTSUP` = `58`
+
+
+#### `wasi_fd_id_t`
+
+- `STDOUT` = `1`
+- `STDERR` = `2`
+
+
+#### `wasi_clock_id_t`
+
+- `REALTIME` = `0`
+- `MONOTONIC` = `1`
+
+
+# Serialization
+
+## HTTP fields and gRPC metadata
+
+TODO
+
+
+[integration]: #Integration
+[memory management]: #Memory-management
+[serialized]: #Serialization
+
+[`proxy_abi_version_0_2_1`]: #proxy_abi_version_0_2_1
+[`_initialize`]: #_initialize
+[`main`]: #main
+[`_start`]: #_start
+[`proxy_on_memory_allocate`]: #proxy_on_memory_allocate
+[`malloc`]: #malloc
+[`proxy_on_context_create`]: #proxy_on_context_create
+[`proxy_on_done`]: #proxy_on_done
+[`proxy_on_log`]: #proxy_on_log
+[`proxy_on_delete`]: #proxy_on_delete
+[`proxy_done`]: #proxy_done
+[`proxy_set_effective_context`]: #proxy_set_effective_context
+[`proxy_on_vm_start`]: #proxy_on_vm_start
+[`proxy_on_configure`]: #proxy_on_configure
+[`proxy_log`]: #proxy_log
+[`proxy_get_log_level`]: #proxy_get_log_level
+[`proxy_get_current_time_nanoseconds`]: #proxy_get_current_time_nanoseconds
+[`proxy_set_tick_period_milliseconds`]: #proxy_set_tick_period_milliseconds
+[`proxy_on_tick`]: #proxy_on_tick
+[`proxy_set_buffer_bytes`]: #proxy_set_buffer_bytes
+[`proxy_get_buffer_bytes`]: #proxy_get_buffer_bytes
+[`proxy_get_buffer_status`]: #proxy_get_buffer_status
+[`proxy_get_header_map_size`]: #proxy_get_header_map_size
+[`proxy_get_header_map_pairs`]: #proxy_get_header_map_pairs
+[`proxy_set_header_map_pairs`]: #proxy_set_header_map_pairs
+[`proxy_get_header_map_value`]: #proxy_get_header_map_value
+[`proxy_add_header_map_value`]: #proxy_add_header_map_value
+[`proxy_replace_header_map_value`]: #proxy_replace_header_map_value
+[`proxy_remove_header_map_value`]: #proxy_remove_header_map_value
+[`proxy_continue_stream`]: #proxy_continue_stream
+[`proxy_close_stream`]: #proxy_close_stream
+[`proxy_on_new_connection`]: #proxy_on_new_connection
+[`proxy_on_downstream_data`]: #proxy_on_downstream_data
+[`proxy_on_downstream_connection_close`]: #proxy_on_downstream_connection_close
+[`proxy_on_upstream_data`]: #proxy_on_upstream_data
+[`proxy_on_upstream_connection_close`]: #proxy_on_upstream_connection_close
+[`proxy_on_request_headers`]: #proxy_on_request_headers
```

Comment:
Should proxy_on_request_metadata and proxy_on_response_metadata also be documented?

Reply:
I've added them while drafting this document, but then I realized that we don't have any access to metadata elements, so those callbacks are pretty much useless, and I've decided not to document them (since they are optional anyway).

Reply:
SGTM, thanks!